src/Workspaces/MSBuild/Core/MSBuild/BuildHostProcessManager.cs

```diff
@@ -63,58 +65,98 @@ public async Task<RemoteBuildHost> GetBuildHostWithFallbackAsync(string projectF
+                    var processStartInfo = CreateBuildHostStartInfo(buildHostKind, pipeName, dotnetPath);
+
+                    var process = Process.Start(processStartInfo);
+                    Contract.ThrowIfNull(process, "Process.Start failed to launch a process.");
+
+                    buildHostProcess = new BuildHostProcess(process, pipeName, _loggerFactory);
+                    buildHostProcess.Disconnected += BuildHostProcess_Disconnected;
+
+                    // We've subscribed to Disconnected, but if the process crashed before that point we might have not seen it
+                    if (process.HasExited)
+                    {
+                        buildHostProcess.LogProcessFailure();
+                        throw new Exception($"BuildHost process exited immediately with {process.ExitCode}");
+                    }
+
+                    // When running on .NET Core, we need to find the right SDK location that can load our project and restart the BuildHost if required.
+                    // When dotnetPath is null, the BuildHost is started with the default dotnet executable, which may not be the right one for the project.
+                    if (buildHostKind == BuildHostProcessKind.NetCore
+                        && projectOrSolutionFilePath is not null
+                        && dotnetPath is null)
+                    {
+                        // The BuildHost will be able to search through all the SDK install locations for a usable MSBuild instance.
+                        var msbuildLocation = await buildHostProcess.BuildHost.FindUsableMSBuildAsync(projectOrSolutionFilePath, cancellationToken).ConfigureAwait(false);
+                        if (msbuildLocation is not null && GetProcessPath() is { } processPath)
+                        {
+                            // The layout of the SDK is such that the dotnet executable is always at the same relative path from the MSBuild location.
+                            dotnetPath = Path.GetFullPath(Path.Combine(msbuildLocation.Path, $"../../{DotnetExecutable}"));
+                            if (dotnetPath is not null && processPath != dotnetPath)
+                            {
+                                // We need to relaunch the .NET BuildHost from a different dotnet instance.
+                                reload = true;
+                                await buildHostProcess.DisposeAsync().ConfigureAwait(false);
+                                _logger?.LogInformation($".NET BuildHost started from {processPath} reloading to start from {dotnetPath} to match necessary SDK location.");
+                            }
+                        }
+                    }
+                } while (reload);
 
                 _processes.Add(buildHostKind, buildHostProcess);
             }
 
             return buildHostProcess.BuildHost;
         }
+
+#if NET
+        static string? GetProcessPath() => Environment.ProcessPath;
```

Comment:
How can this return null?

Reply:
> If the executable is renamed or deleted before this property is first accessed, the return value is undefined and depends on the operating system.

https://learn.microsoft.com/en-us/dotnet/api/system.environment.processpath?view=net-9.0#remarks